core/src/main/java/org/apache/carbondata/core/datastore/block/SegmentPropertiesAndSchemaHolder.java

```diff
@@ -332,13 +334,42 @@ public void clear() {
       }
       SegmentPropertiesAndSchemaHolder.SegmentPropertiesWrapper other =
           (SegmentPropertiesAndSchemaHolder.SegmentPropertiesWrapper) obj;
-      return tableIdentifier.equals(other.tableIdentifier) && columnsInTable
-          .equals(other.columnsInTable) && Arrays
+      return tableIdentifier.equals(other.tableIdentifier) && checkColumnSchemaEquality(
+          columnsInTable, other.columnsInTable) && Arrays
           .equals(columnCardinality, other.columnCardinality);
     }
 
+    private boolean checkColumnSchemaEquality(List<ColumnSchema> obj1, List<ColumnSchema> obj2) {
+      List<ColumnSchema> clonedObj1 = new ArrayList<>(obj1);
```

Comment:
You can add the first check for length in the first line of method...if length of 2 lists is not same then we can return false from here itself

Reply:
I think checkColumnSchemaEquality method need consider "obj2 == null"